From f8e682d647776942391655f7fa118984a123f221 Mon Sep 17 00:00:00 2001
From: David Daney <david.daney@cavium.com>
Date: Fri, 24 May 2019 22:25:20 +0200
Subject: [PATCH 207/345] MIPS: OCTEON: Print warning message if OCTEON II
 kernel run on earlier chips.

If the kernel is configured with CONFIG_CAVIUM_OCTEON2, it will
contain machine instructions that cause ReservedInstruction exceptions
on earlier chips.  Print a nice warning message rather than crashing.

Signed-off-by: David Daney <david.daney@cavium.com>
---
 arch/mips/cavium-octeon/setup.c               |  4 +++
 .../mach-cavium-octeon/kernel-entry-init.h    | 27 +++++++++++++++++++
 2 files changed, 31 insertions(+)

diff --git a/arch/mips/cavium-octeon/setup.c b/arch/mips/cavium-octeon/setup.c
index 4c8818c2e7b7..04864fe42bea 100644
--- a/arch/mips/cavium-octeon/setup.c
+++ b/arch/mips/cavium-octeon/setup.c
@@ -91,6 +91,10 @@ struct octeon_boot_descriptor *octeon_boot_desc_ptr;
 struct cvmx_bootinfo *octeon_bootinfo;
 EXPORT_SYMBOL(octeon_bootinfo);
 
+const char octeon_not_compatible[] =
+	"ERROR: CONFIG_CAVIUM_OCTEON2 not compatible with this processor\r\n"
+	"You must rebuild the kernel to be able to use it on this system.\r\n";
+
 #ifdef CONFIG_KEXEC
 #ifdef CONFIG_SMP
 /*
diff --git a/arch/mips/include/asm/mach-cavium-octeon/kernel-entry-init.h b/arch/mips/include/asm/mach-cavium-octeon/kernel-entry-init.h
index 5294efb3f5fb..9e4a095cc06a 100644
--- a/arch/mips/include/asm/mach-cavium-octeon/kernel-entry-init.h
+++ b/arch/mips/include/asm/mach-cavium-octeon/kernel-entry-init.h
@@ -290,6 +290,33 @@ FEXPORT(octeon_hotplug_entry)
 octeon_main_processor:
 	dla	v0, octeon_cvmseg_lines
 	sw	t2, 0(v0)
+#ifdef CONFIG_CAVIUM_OCTEON2
+	mfc0	v1, CP0_PRID_REG
+	andi	v1, 0xff00
+	srl	v1, 8
+	addiu	v1, -0x90
+	bgez	v1, 1f
+	/* Pre-OCTEON II running with CONFIG_CAVIUM_OCTEON2 will not work. */
+	ld	v1, 0x140(a3)
+	dli	v0, 0x8001180000000800 /* UART base */
+	bbit0	v1, 35, 2f /* Check for UART1 */
+	daddiu	v0, 0x400
+2:
+	dla	v1, octeon_not_compatible
+3:
+	lbu	t0, 0(v1)
+	daddiu	v1, 1
+	beqz	t0, 4f
+5:
+	ld	t1, 0x28(v0) /* LSR*/
+	bbit0	t1, 5, 5b
+	sd	t0, 0x40(v0)
+	b	3b
+4:
+	wait
+	b	4b
+1:
+#endif /* CONFIG_CAVIUM_OCTEON2 */
 	.set pop
 .endm
 
-- 
2.25.1

